lib/vanagon/project.rb

```diff
@@ -110,18 +110,81 @@ def get_provides
       provides.flatten.uniq
     end
 
-    # Collects the pre-install actions for the project and it's components
+    # Collects the preinstall packaging actions for the project and it's components
+    # for the specified packaging state
     #
-    # @return [Array] array of Bourne shell compatible scriptlets to execute
-    def get_preinstall_actions
-      @components.map(&:preinstall_actions).flatten
+    # @param pkg_state [String] the package state we want to run the given scripts for.
+    #   Can be one or more of 'install' or 'upgrade'
+    # @return [Array] array of Bourne shell compatible scriptlets to execute during the preinstall
+    #   phase of packaging during the state of the system defined by pkg_state (either install or upgrade)
+    def get_preinstall_actions(pkg_state)
+      scripts = []
+      @components.map(&:preinstall_actions).flatten.compact.select { |s| s.pkg_state.include? pkg_state }.each do |action|
```

Comment:
Also, i think this can just be:
`@components.map(&:preinstall_actions).flatten.compact.select { |s| s.pkg_state.include? pkg_state }.map(&:action).join("\n")`